src/Components/ADT3DSceneBuilder/Internal/VisualRuleForm/Internal/ConditionsCallout/ConditionsCallout.tsx

```diff
@@ -0,0 +1,255 @@
+    Stack,
+    DirectionalHint,
+    TextField,
+    PrimaryButton,
+    DefaultButton
+} from '@fluentui/react';
+import { ConditionSummary } from './ConditionSummary';
+import { ActionItem } from './ActionItem';
+import { ConditionCalloutReducer } from './ConditionsCallout.state';
+import { IDTDLPropertyType } from '../../../../../../Models/Types/Generated/3DScenesConfiguration-v1.0.0';
+import { useTranslation } from 'react-i18next';
+import { areRangesValid, checkValidity } from './ConditionCalloutUtility';
+import { deepCopy } from '../../../../../../Models/Services/Utils';
+import { CalloutInfoType } from '../ConditionsList.types';
+
+const getClassNames = classNamesFunction<
+    IConditionsCalloutStyleProps,
+    IConditionsCalloutStyles
+>();
+
+const ROOT_LOC = '3dSceneBuilder.visualRuleForm';
+const LOC_KEYS = {
+    addCondition: `${ROOT_LOC}.addCondition`,
+    editCondition: `${ROOT_LOC}.editCondition`,
+    labelField: `${ROOT_LOC}.labelField`
+};
+
+const ConditionsCallout: React.FC<IConditionsCalloutProps> = (props) => {
+    const {
+        calloutType,
+        isOpen,
+        onDismiss,
+        onSave,
+        target,
+        styles,
+        valueRange,
+        valueRangeType
+    } = props;
+
+    // state
+
+    // refs
+    const validityMap = useRef<ConditionValidityMap>({
+        label: !!valueRange.visual.labelExpression?.length,
+        ranges: areRangesValid(valueRange.values, valueRangeType)
+    });
+
+    // hooks
+    const { t } = useTranslation();
+    const [
+        conditionCalloutState,
+        conditionCalloutDispatch
+    ] = useReducer<ConditionCalloutReducerType>(ConditionCalloutReducer, {
+        conditionToEdit: valueRange,
+        originalCondition: valueRange,
+        isDirty: false
+    });
+
+    // side-effects
+    useEffect(() => {
+        // RESET CONDITION IN STATE AND VALIDITY MAP
+        conditionCalloutDispatch({
+            type: ConditionCalloutActionType.RESET_CONDITION,
+            payload: { valueRange: valueRange }
+        });
+        validityMap.current = {
+            label: !!valueRange.visual.labelExpression?.length,
+            ranges: areRangesValid(valueRange.values, valueRangeType)
+        };
+    }, [valueRange, valueRangeType]);
+
+    // callbacks
+    const onLabelChange = useCallback(
+        (
+            _event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
+            newValue?: string
+        ) => {
+            if (newValue) {
+                validityMap.current.label = true;
+            } else {
+                validityMap.current.label = false;
+            }
+            conditionCalloutDispatch({
+                type: ConditionCalloutActionType.FORM_CONDITION_LABEL_SET,
+                payload: { label: newValue }
+            });
+        },
+        [conditionCalloutDispatch]
+    );
+
+    const onValuesChange = useCallback(
+        (
+            valueType: IDTDLPropertyType,
+            newValues: unknown[],
```

Comment:
why don't we know the type of the values? surely it's at least string[] or number[]?

Reply:
String, number or boolean. I could narrow it down to those 3

Reply:
I'd have to cast it though, since the type is unknown[ ] in our types file.

Reply:
I'd rather do a caste at one point than have unknown spread throughout the codebase. I'm finding all sorts of crazy bugs and issues in OAT because of their use of any and unknowns.